Arch considerations upgrading (#919)
digital-asset/docs.daml.com: docs/3.1/docs/tsa-arch/arch-considerations-upgrading.rst

3.1.8.1 Understanding Contract Archival in Daml
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
At a purely technical level, "archived" means a consuming choice has been exercised on the contract. However, there are meaningful distinctions between the various business and operational contexts in which an archive may occur.

1. Natural End of Lifecycle: The contract represents a business entity whose lifecycle has naturally ended. For example, a loan contract might be archived when the loan is fully repaid.

2. State No Longer Holds True: The contract attested to a certain business state, but that state is no longer valid. For instance, an agreement that was conditional on certain criteria might be archived when those criteria are no longer met.

3. Modification of the Underlying Entity/State: The business entity or state represented by the contract itself is still relevant; however, because Daml contracts are immutable, the update requires archiving the outdated contract. If the updated contract is written using v2, this results in an organic and incremental migration away from v1.

> 4. Explicit Upgrade: The contract is archived as part of an upgrade process, preferably by an upgrade runner. This can be done during planned downtime, but in most cases, it can be done incrementally via a throttled background process. As a business operation, this is distinct from the previous three.

Reviewer: Is it worth unpacking the why of "preferably by an upgrade runner" briefly here?

> While all of these cases result in an "archive" ledger event, they represent different business operations with varying levels of impact. The ordering (1 → 4) reflects a general preference for minimal impact, with natural expiration being the least disruptive and explicit upgrades requiring more intervention.

Reviewer: Move this paragraph to before the list so the reason for the ordering is clear.

3.2 Backward Compatibility in Backend Code
------------------------------------------
To ensure that ledger reads in the v2 backend remain compatible with contracts created using v1 Daml models, use transaction and contract filters with symbolic package references. These references take the form of `#package-name:module-name:template-id` for ledger reads to retrieve data from all contracts that are instances of the template with `module-name` and `template-id` of any version of the `package-name`. 

Since newer versions of a template may introduce fields of type `Optional` that did not exist in earlier versions, the backend must handle cases where these fields are missing. The Daml SDK’s codegens assist with this by automatically setting missing `Optional` fields to `None`.

3.3 Manage Backward-Incompatible Changes
----------------------------------------

> Not all changes can maintain backward compatibility. The strategy used for updating Daml models follows similar principles of how APIs evolve in a service-based architecture. 

Reviewer: ```suggestion
Not all changes can maintain backward compatibility. The strategy for updating Daml models follows principles similar to how APIs evolve in a service-based architecture. 
```

Avoid package name conflicts, particularly between packages published by different app providers. Follow the Java ecosystem’s convention of prefixing package names with the reverse Domain Name System (DNS) name of the app provider. For example, for the issuance workflows of the money market fund app provided by Acme Inc., the recommended `daml.yaml` configuration would be: `name: com-acme-money-market-fund-issuance`.

4. Package Vetting, Testing, and Rollback
=========================================

4.1 Package Vetting
-------------------

4.1.1 Vetting
~~~~~~~~~~~~~
By default, when a Daml package is uploaded, the participant node automatically marks it as vetted and publishes its vetting status on the synchronizer. This allows other participant nodes to determine which workflows the parties on the participant can engage in. A package cannot be used until it is vetted, providing an additional verification step in the deployment process.

4.1.2 Unvetting
~~~~~~~~~~~~~~~
Packages can also be unvetted. For example, after uploading and vetting v2, unvetting v1 signals that the participant node can no longer participate in v1 workflows, finalizing the upgrade process. 

> Note that all v1 contracts must be fully upgraded before unvetting v1 to avoid potential issues. This requirement extends beyond a zero-downtime upgrade.

Reviewer 2: This sentence can create confusion. Smart contract upgrading does not require any explicit contract upgrade prior to their use in upgraded workflows (i.e. a contract created with a v1 version of a template can be used in v2 workflows)

Author: Hi Tudor, thank you for reviewing my work. Could you please confirm if replacing the "Note that all ..." part with the following conveys all the necessary clarity and is technically accurate?

---

Note that smart contract upgrading does not require any explicit contract upgrade before use in upgraded workflows. While v1 contracts can continue to be used in v2 workflows, unvetting v1 ensures that v1 workflows are unavailable. Since existing v1 contracts can remain active if they are compatible with v2 workflows, all v1 contracts must either be confirmed as compatible with v2 workflows or fully migrated to v2, and finalized in any active v1 workflows, if any exist, before unvetting v1 to prevent issues. This requirement extends beyond a zero-downtime upgrade, ensuring system consistency and preventing errors during the upgrade process.

* A workflow-level compatibility test verifies that core business processes (workflows) continue to function correctly after an upgrade. At a minimum, it is recommended to include one integration test. A basic integration test should follow these steps:

  1. Start the application with v2 software, but upload only the v1 DAR file to test backward compatibility.
  2. Initialize the application and start one instance of every core workflow.
  3. Upload the v2 DAR.
  4. Update the configuration to instruct the backends to start using the v2 DAR.
  5. Verify that the workflows remain in the correct state and can continue without issues. 

For more complex upgrades, additional tests may be needed. 

4.3 Rollback
------------

4.3.1 Rollback by Unvetting
~~~~~~~~~~~~~~~~~~~~~~~~~~~
To roll back upgrades that do not modify the types of existing templates and choices, unvet the v2 DAR package.

4.3.2 Rollback by Roll-Forward
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> For upgrades that add new fields to existing templates, rollback becomes more complex. In such cases, the rollback must be performed in a “roll-forward” fashion by publishing a new upgrade. This is necessary because if at least one contract has been created using the new fields, those contracts cannot be read with the previous version of the Daml code. Simply unvetting v2 is not an option unless it is acceptable for contracts using the new fields to no longer be referenced. Instead, follow these steps:

Reviewer: ```suggestion
Rollback is more complex for upgrades that add new fields to existing templates. In such cases, the rollback must be performed as a “roll-forward” by publishing a new upgrade. This is necessary because if at least one contract has been created using the new fields, those contracts cannot be read with the previous version of the Daml code. Unvetting v2 is not an option unless it is acceptable for contracts using the new fields to no longer be referenced. Instead, follow these steps:
```

1. Publish a new version of the DARs that disregards the newly added fields.
2. Introduce a `Downgrade` choice in the new version that resets the newly added fields to `None`, making the contracts compatible with the original version.
3. Use backend automation to iterate through the ACS and invoke the `Downgrade` choice.

To avoid complex "roll-forward" rollbacks, consider breaking an upgrade that introduces new fields into two steps:

> 1. Introduce an upgrade that adds the new fields but does not use them. Since no changes are made to the choices, this upgrade will not require a rollback in case of a bug.

Reviewer: ```suggestion
1. Introduce an upgrade that adds the new fields but does not use them. Since no changes are made to the choices, this upgrade does not require a rollback in case of a bug.
```
